pkg/storage/mysql/mysqlstore.go

```diff
@@ -0,0 +1,271 @@
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	// Add as per the documentation - https://github.com/go-sql-driver/mysql
+	_ "github.com/go-sql-driver/mysql"
+
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+)
+
+// Provider represents an MySQL DB implementation of the storage.Provider interface
+type Provider struct {
+	dbPath string
+	db     *sql.DB
+	dbs    map[string]*sqlDBStore
+	sync.RWMutex
+}
+
+type sqlDBStore struct {
+	db     *sql.DB
+	dbName string
+}
+
+const (
+	blankDBPathErrMsg         = "DB URL for new mySQL DB provider can't be blank"
+	failToCloseProviderErrMsg = "failed to close provider"
+	//nolint:varcheck deadcode
+	tablePrefix   = "t_"
+	sqlDBNotFound = "no rows"
+)
+
+type result struct {
+	key   string
+	value []byte
+}
+
+// NewProvider instantiates Provider
+func NewProvider(dbPath string) (*Provider, error) {
+	if dbPath == "" {
+		return nil, errors.New(blankDBPathErrMsg)
+	}
+
+	// Example DB Path root:my-secret-pw@tcp(127.0.0.1:3306)/
+	db, err := sql.Open("mysql", dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open connection: %w", err)
+	}
+
+	p := &Provider{
+		dbPath: dbPath,
+		db:     db,
+		dbs:    map[string]*sqlDBStore{}}
+
+	return p, nil
+}
+
+// OpenStore opens and returns new db for given name space.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	// creating the database
+	_, err := p.db.Exec("CREATE DATABASE IF NOT EXISTS `name`")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create db: %w", err)
+	}
+
+	// Use is used to select the created database without this DDL operations are not permitted
+	_, err = p.db.Exec("USE `name`")
+	if err != nil {
+		return nil, fmt.Errorf("failed to use db: %w", err)
+	}
+
+	store := &sqlDBStore{
+		db:     p.db,
+		dbName: name}
+
+	p.dbs[name] = store
+
+	return store, nil
+}
+
+// Close closes the provider.
+func (p *Provider) Close() error {
+	p.Lock()
+	defer p.Unlock()
+
+	for _, store := range p.dbs {
+		err := store.db.Close()
+		if err != nil {
+			return fmt.Errorf(failToCloseProviderErrMsg+": %w", err)
+		}
+	}
+
+	if err := p.db.Close(); err != nil {
+		return err
+	}
+
+	p.dbs = make(map[string]*sqlDBStore)
+
+	return nil
+}
+
+// CloseStore closes a previously opened store
+func (p *Provider) CloseStore(name string) error {
+	p.Lock()
+	defer p.Unlock()
+
+	store, exists := p.dbs[name]
+	if !exists {
+		return nil
+	}
+
+	delete(p.dbs, name)
+
+	conn, err := store.db.Conn(context.Background())
+	if err != nil {
+		return err
+	}
+
+	return conn.Close()
+}
+
+// Put stores the key and the record
+func (s *sqlDBStore) Put(k string, v []byte) error {
+	if k == "" || v == nil {
+		return errors.New("key and value are mandatory")
+	}
+
+	createTableStmt := "CREATE Table IF NOT EXISTS `tablePrefix s.dbName` (`key` varchar(50) NOT NULL, `value` BLOB, " +
```

Comment:
this should've been done when instantiating `sqlDBStore`, not on every `Put()`

Reply:
also, is `varchar(50)` enough width?

Reply:
MySQL will complain if you try to create indexes that are wider than its configured maximum.

Instead, control the width of the key inserted into the database by hashing that. Set the width of the varchar appropriately.

Reply:
... which reminds me, I don't see the creation of an index on column `key`

Reply:
This PR is getting bigger, I have increased the varchar(50) to varchar(255). Created TODO #1940. Which i will update right after this PR.

Reply:
Actually, I realized storing the hash would break the `Iterator()` function. So, don't use the hash. But we still need a way to implement indexes on keys of arbitrary lengths.

Reply:
Can i address that in the separate PR. Will need to dig deeper how to get achieve that

Reply:
@talltree sure - can leave a TODO